synapse/federation/sender/per_destination_queue.py

```diff
@@ -204,97 +206,117 @@ def _transaction_transmission_loop(self):
 
                 pending_edus = device_update_edus + to_device_edus
 
-                # BEGIN CRITICAL SECTION
-                #
-                # In order to avoid a race condition, we need to make sure that
-                # the following code (from popping the queues up to the point
-                # where we decide if we actually have any pending messages) is
-                # atomic - otherwise new PDUs or EDUs might arrive in the
-                # meantime, but not get sent because we hold the
-                # transmission_loop_running flag.
-
-                pending_pdus = self._pending_pdus
-
-                # We can only include at most 50 PDUs per transactions
-                pending_pdus, self._pending_pdus = pending_pdus[:50], pending_pdus[50:]
-
-                pending_edus.extend(self._get_rr_edus(force_flush=False))
-                pending_presence = self._pending_presence
-                self._pending_presence = {}
-                if pending_presence:
-                    pending_edus.append(
-                        Edu(
-                            origin=self._server_name,
-                            destination=self._destination,
-                            edu_type="m.presence",
-                            content={
-                                "push": [
-                                    format_user_presence_state(
-                                        presence, self._clock.time_msec()
-                                    )
-                                    for presence in pending_presence.values()
-                                ]
-                            },
+                # Make a transaction sending opentracing span, this span follows on from
+                # all the  edus in that transaction. This needs to be done since there is
+                # no active span here and if the edus are not received by the remote the
+                # span would have no causality and it would be forgotten.
+
+                span_contexts = [
+                    extract_text_map(
+                        json.loads(
+                            edu.get_dict().get("content", {}).get("context", "{}")
                         )
                     )
-
-                pending_edus.extend(
-                    self._pop_pending_edus(MAX_EDUS_PER_TRANSACTION - len(pending_edus))
-                )
-                while (
-                    len(pending_edus) < MAX_EDUS_PER_TRANSACTION
-                    and self._pending_edus_keyed
-                ):
-                    _, val = self._pending_edus_keyed.popitem()
-                    pending_edus.append(val)
-
-                if pending_pdus:
-                    logger.debug(
-                        "TX [%s] len(pending_pdus_by_dest[dest]) = %d",
-                        self._destination,
-                        len(pending_pdus),
+                    for edu in pending_edus
```

Comment:
`pending_edus` at this point doesn't include all the EDUs that will be sent. Should we just wrap the send_new_transaction bit?

Reply:
Ah fair, I've moved the opentracing from this file to `transaction_manager.py`